PreconditionChebyshev: report stats with output_details

tjhei opened the pull request:
I would like to get some information about what eigenvalue estimates are computed in PreconditionChebyshev and how many CG iterations we needed. Using deallog seemed the natural thing to do.
Note: this will break several tests as this output is recorded in the tests. I can fix the output, but the results might be slightly machine-specific. Any thoughts on how to proceed?


FYI: @tcclevenger

peterrum commented:
@tjhei what about introducing an additional flag in the AdditionalDara and setting it to false per default?

tjhei commented:
> @tjhei what about introducing an additional flag in the AdditionalDara and setting it to false per default?

I like it. We do this in other places as well, I think. ``TrilinosWrappers::AdditionalData::output_details`` for example.

tjhei commented:
Maybe something like this? To mirror ``output_details`` in other preconditioners, I am writing to ``std::cout`` as well, but I don't feel strongly about it (having to enable the deallog would be annoying).

bangerth commented:
We've moved away from these static ways of exporting information. The linear solvers have all received signals to which one can attach oneself. I think that that would also be the right approach for preconditioners.

tjhei commented:
> We've moved away from these static ways of exporting information.

I changed it to the way it is done in TrilinosWrappers::PreconditionAMG (flag output_details and printing to std::cout). If you feel strongly about it, I can make a callback instead. It means a little bit more code on the user side, but obviously it is more flexible.

kronbichler commented:
> I changed it to the way it is done in TrilinosWrappers::PreconditionAMG (flag output_details and printing to std::cout).

How about using the flag as it is right now and open an issue then? While I definitely agree that we should transition towards signals, I think we should first go forward with this topic (unless @tjhei has free resources to do it immediately here).

bangerth commented:
I'm not excited about writing to `std::cout`. There are actually only a handful of places in the library where we do that, and almost all of them are guarded by `#if DEBUG` or something similar. We shouldn't introduce an approach that we've been strenuously tried to eradicate just so we can replace it again in a follow-up patch. Why not do it right right away?

(As a complete aside, the xSDK guidelines also strongly recommend not writing to `cout`.)

bangerth commented:
I'll add that you'll also get jumbled output if you do this with a parallel vector because multiple processors will all try to write at once.

tjhei commented:
Alright, I will report back after I get to this.

tjhei commented:
Updated. I decided to return a struct with the information instead of adding a ton of complication for a callback. Does that make sense?
One question: I don't like how complicated the fully qualified type is, as it is nested inside PreconditionChebychev. The user will have to use ``auto`` or write this long ``typename PreconditionChebychev<...`` . How do you all feel about moving this struct outside the class?

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -1160,6 +1160,30 @@ class PreconditionChebyshev : public Subscriptor
   n() const;
 
+  /**
+   * Struct that contains information about the eigenvalue estimation performed
```
```suggestion
   * A struct that contains information about the eigenvalue estimation performed
```

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -1160,6 +1160,30 @@ class PreconditionChebyshev : public Subscriptor
 
+  /**
+   * Struct that contains information about the eigenvalue estimation performed
+   * by this class.
```
"this class" is ambiguous (the current class? the surrounding class?)
```suggestion
   * by the (surrounding PreconditionChebychev) class.
```

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -1160,6 +1160,30 @@ class PreconditionChebyshev : public Subscriptor
+    /**
+     * Estimate for the smallest eigenvalue.
+     */
+    double min_eigenvalue;
```
Would you mind reflecting in the name that these are *estimates*? E.g.
```suggestion
    double min_eigenvalue_estimate;
```
or 
```suggestion
    double min_estimated_eigenvalue;
```

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -1160,6 +1160,30 @@ class PreconditionChebyshev : public Subscriptor
+    /**
+     * Estimate for the largest eigenvalue.
+     */
+    double max_eigenvalue;
```
same here

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -1160,6 +1160,30 @@ class PreconditionChebyshev : public Subscriptor
+    /**
+     * The degree (either as set or estimated).
+     */
+    unsigned int degree;
```
The degree of what? (Both in the variable name and in the comment.)

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -2297,25 +2325,28 @@ PreconditionChebyshev<MatrixType, VectorType, PreconditionerType>::
           // include a safety factor since the CG method will in general not
           // be converged
-          max_eigenvalue = 1.2 * eigenvalue_tracker.values.back();
+          info.max_eigenvalue = 1.2 * eigenvalue_tracker.values.back();
```
Should this factor be reflected in the documentation somewhere? It means that even if you run the solver to convergence, you'll get an overestimate by a factor of 1.2...

tjhei commented:
I added some documentation and renamed variables.

bangerth commented:
It currently fails:
```
clang-tidy-6.0 -header-filte/jenkins/workspace/dealii-tidy_PR-9645/include/deal.II/lac/precondition.h:2347:43: error: no member named 'max_eigenvalue_estimate' in 'dealii::PreconditionChebyshev<Step37::LaplaceOperator<3, 2, float>, dealii::LinearAlgebra::distributed::Vector<float, dealii::MemorySpace::Host>, dealii::DiagonalMatrix<dealii::LinearAlgebra::distributed::Vector<float, dealii::MemorySpace::Host> > >::AdditionalData' [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-9645/include/deal.II/lac/precondition.h:2349:14: error: no member named 'max_eigenvalue_estimate' in 'dealii::PreconditionChebyshev<Step37::LaplaceOperator<3, 2, float>, dealii::LinearAlgebra::distributed::Vector<float, dealii::MemorySpace::Host>, dealii::DiagonalMatrix<dealii::LinearAlgebra::distributed::Vector<float, dealii::MemorySpace::Host> > >::AdditionalData' [clang-diagnostic-error]

clang-tidy-6.0 -header-fil/jenkins/workspace/dealii-tidy_PR-9645/include/deal.II/lac/precondition.h:2347:43: error: no member named 'max_eigenvalue_estimate' in 'dealii::PreconditionChebyshev<Step59::LaplaceOperator<3, 8, float>, dealii::LinearAlgebra::distributed::Vector<float, dealii::MemorySpace::Host>, Step59::PreconditionBlockJacobi<3, 8, float> >::AdditionalData' [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-9645/include/deal.II/lac/precondition.h:2349:14: error: no member named 'max_eigenvalue_estimate' in 'dealii::PreconditionChebyshev<Step59::LaplaceOperator<3, 8, float>, dealii::LinearAlgebra::distributed::Vector<float, dealii::MemorySpace::Host>, Step59::PreconditionBlockJacobi<3, 8, float> >::AdditionalData' [clang-diagnostic-error]
```

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -1160,6 +1164,31 @@ class PreconditionChebyshev : public Subscriptor
+    unsigned int cg_iterations;
+    /**
+     * The polynomial degree of the Chebyshev (either as set using
+     * AdditionalData::degree or estimated as described there).
```
...of the Chebyshev WHAT? I think there is a word missing here.

bangerth commented:
Do you also want to add a changelog entry?
